doc/coq.rst

```diff
@@ -0,0 +1,383 @@
+.. _coq:
+
+***
+Coq
+***
+
+.. contents:: Table of Contents
+    :depth: 3
+
+Introduction
+------------
+
+Dune is able to build Coq theories and plugins with support for extraction and
+.mlg file preprocessing.
+
+A *Coq theory* is a collection of ``.v`` files, perhaps in several
+subdirectories, all prefixed by a common Coq module name, following the
+directory hierarchy for module names.
+
+A *Coq plugin* is an OCaml :ref:`library`, that is intended to be dynamically
+loaded by Coq at runtime when compiling a Coq theory, and is typically linked
+with the Coq OCaml API.
+
+We may indistinctively refer to a :ref:`dune-project` containing Coq theories as
+a *Coq project*.
+
+A *Coq project* is a collection of Coq plugins and Coq theories. This notion is
+subsumed by that of a :ref:`dune-project`, and should not be considered a
+technical definition.
+
+The ``.v`` files of a theory do not only have to be present as source files.
+They may also be created as Dune targets from other rules.
+
+.. _coq-theory:
+
+coq.theory
+----------
+
+To enable Coq support in a Dune project. The Coq language version should be
+selected in the :ref:`dune-project` file. For example:
+
+.. code:: scheme
+
+    (using coq 0.3)
+
+This will allow for the ``coq.theory`` stanza and other ``coq.*`` stanzas to be
+used in the current project. See the :ref:`Dune Coq language section<coq-lang>`
+for more details.
+
+The Coq theory stanza is very similar in form to the OCaml :ref:`library`
+stanza:
+
+.. code:: scheme
+
+    (coq.theory
+     (name <module_prefix>)
+     (package <package>)
+     (synopsis <text>)
+     (modules <ordered_set_lang>)
+     (libraries <ocaml_libraries>)
+     (flags <coq_flags>)
+     (mode <coq_native_mode>)
+     (theories <coq_theories>))
+
+The stanza will build all ``.v`` files in the given directory and also all
+subdirs if the :ref:`include-subdirs <include-subdirs-coq>` stanza is present.
+The semantics of fields are:
+
+- ``<module_prefix>`` is a dot-separated list of valid Coq module names and
+  determines the module scope under which the theory is compiled (this
+  corresponds to Coq's ``-R`` option).
+  
+  For example, if ``<module_prefix>`` is ``foo.Bar``, the theory modules will be
+  named as ``foo.Bar.module1``, ``foo.Bar.module2``, etc. Note that modules in
+  the same theory don't see the ``foo.Bar`` prefix in the same way that OCaml
+  ``wrapped`` libraries do.
+  
+  For compatibility reasons, Coq lang 1.0 will install a theory named
```

Comment:
Maybe "Coq language version 1.0" and hyperlink to "(using coq 0.3)" above.

Reply:
Coq lang 1.0 is a hypothetical "out of experimentation" version that Emilio came up with intending to describe features that will stabilize. I think I will link to the Coq Dune lang section, and describe there, a little more, that hypothetical.